pkg/trace/api/api.go

```diff
@@ -562,10 +561,49 @@ func (r *HTTPReceiver) Languages() string {
 	return strings.Join(str, "|")
 }
 
-func decodeRequest(req *http.Request, dest msgp.Decodable) error {
+//var bufioReaderPool sync.Pool
+//
+//func newBufioReader(r io.Reader) *bufio.Reader {
+//	if v := bufioReaderPool.Get(); v != nil {
+//		br := v.(*bufio.Reader)
+//		br.Reset(r)
+//		return br
+//	}
+//
+//	// Note: if this reader size is ever changed, update
+//	// TestHandlerBodyClose's assumptions.
+//	//return bufio.NewReaderSize(r, 1048576)
+//	//return bufio.NewReaderSize(r, 5242880)
+//}
+//
+//func putBufioReader(br *bufio.Reader) {
+//	br.Reset(nil)
+//	bufioReaderPool.Put(br)
+//}
+
+func decodeRequest(req *http.Request, dest *pb.Traces) error {
+	//br := newBufioReader(req.Body)
+	//defer putBufioReader(br)
 	switch mediaType := getMediaType(req); mediaType {
 	case "application/msgpack":
-		return msgp.Decode(req.Body, dest)
+		if req.ContentLength > 0 {
+			bs := make([]byte, req.ContentLength)
+			_, err := io.ReadFull(req.Body, bs)
+			if err != nil {
+				return err
+			}
+			return directDecodeTraces(bs, dest)
+		}
+		panic("failed to decode request.")
+		// Normally we don't use ioutil.ReadAll, but req.Body is limited to maxRequestBodyLength
+		// with a LimitedReader.
+		bs, err := ioutil.ReadAll(req.Body)
```

Comment:
This is less optimal, but can be greatly improved from `ioutil.ReadAll`. `ioutil.ReadAll` uses a `bytes.Buffer`, which does not grow itself optimally for this use case. That can be optimized fairly easily.